### ISSUE TITLE
container_memory_working_set_bytes shows previous container memory

### ISSUE THREAD

**Issue by froblesmartin** [kubernetes/kubernetes#128538]
### What happened?

As raised at https://github.com/prometheus-operator/kube-prometheus/issues/2522, the `container_memory_working_set_bytes` metric shows memory from a killed container instance which is not running anymore, not allowing to use it to know the "real" memory usage.

In my case, I still see the last memory used by the previous container instance for 4:30 minutes, but from @vladmalynych I see in his case it seems to be only for around 3 minutes.

### What did you expect to happen?

To only show data from the **current** running container.

### How can we reproduce it (as minimally and precisely as possible)?

Trigger a container restart (OOMKilled or Eviction) and check the resulting metric.

### Kubernetes version

<details>

```console
Client Version: v1.31.2
Kustomize Version: v5.4.2
Server Version: v1.31.1-gke.1678000 (Also happening in 1.29.8-gke.1211000)
```

</details>

### Cloud provider

<details>
GKE
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.5 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.5 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
$ uname -a
Linux gke-*** 5.15.0-1067-gke #73-Ubuntu SMP Sat Aug 31 04:29:32 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by rexagod** [comment]
> Trigger a container restart (OOMKilled or Eviction) and check the resulting metric.

I'll try to recreate this but it seems highly unlikely for such a trivial use-case to be faulty, was this reproducible 100% of the time?

**Comment by froblesmartin** [comment]
> > Trigger a container restart (OOMKilled or Eviction) and check the resulting metric.
> 
> I'll try to recreate this but it seems highly unlikely for such a trivial use-case to be faulty, was this reproducible 100% of the time?

Yes, it still happens to us. Did have a look at any running instance to see if it is also the case for you?

**Comment by kyrofa** [comment]
This is happening for me as well, also when OOMKills happen, every time. The restarted container gets its own metric because it has a different name,  and the old one sticks around for a few minutes. My solution is to `min(...) by (pod, container)`.

**Comment by SergeyKanzhelev** [comment]
This looks similar to https://github.com/kubernetes/kubernetes/issues/134049

/sig node

@kyrofa the bug https://github.com/kubernetes/kubernetes/issues/134049 marked as need information as we cannot repro. Can you comment on specific repro steps and exact environment?
